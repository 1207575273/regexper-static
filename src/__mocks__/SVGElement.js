import React from 'react';

import Base from 'components/SVG/Base';

class SVGElement extends Base {
  reflow() {
    return this.setBBox(this.props.bbox);
  }

  render() {
    return <text>Mock content</text>;
  }
}

export default SVGElement;
