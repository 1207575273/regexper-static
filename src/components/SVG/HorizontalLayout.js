import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';

import style from './style';

import reflowable from './reflowable';
import Path from './path';

@reflowable
class HorizontalLayout extends React.PureComponent {
  static defaultProps = {
    withConnectors: false,
    spacing: 10
  }

  state = {
    childTransforms: List()
  }

  updateChildTransforms(childBoxes) {
    return this.state.childTransforms.withMutations(transforms => (
      childBoxes.forEach((box, i) => (
        transforms.set(i, `translate(${ box.offsetX } ${ box.offsetY })`)
      ))
    ));
  }

  updateConnectorPaths(childBoxes) {
    let last = childBoxes[0];

    return childBoxes.slice(1).reduce((path, box) => {
      try {
        return path
          .moveTo({ x: last.offsetX + last.axisX2, y: this.getBBox().axisY })
          .lineTo({ x: box.offsetX + box.axisX1 });
      }
      finally {
        last = box;
      }
    }, new Path()).toString();
  }

  reflow() {
    const { spacing, withConnectors } = this.props;

    const childBoxes = this.children.map(child => child.getBBox());
    const verticalCenter = childBoxes.reduce((center, box) => Math.max(center, box.axisY), 0);
    const width = childBoxes.reduce((width, box) => width + box.width, 0) + (childBoxes.length - 1) * spacing;
    const height = childBoxes.reduce((ascHeight, box) => Math.max(ascHeight, box.axisY), 0) +
                   childBoxes.reduce((decHeight, box) => Math.max(decHeight, box.height - box.axisY), 0);
    this.setBBox({ width, height, axisY: verticalCenter }, { axisX1: true, axisX2: true });

    let offset = 0;
    childBoxes.forEach(box => {
      box.offsetX = offset;
      box.offsetY = this.getBBox().axisY - box.axisY;
      offset += box.width + spacing;
    });

    this.setStateAsync({
      childTransforms: this.updateChildTransforms(childBoxes),
      connectorPaths: withConnectors ? this.updateConnectorPaths(childBoxes) : ''
    });
  }

  childRef = i => child => this.children[i] = child

  render() {
    const { children } = this.props;
    const { childTransforms, connectorPaths } = this.state;

    this.children = [];

    return <React.Fragment>
      <path d={ connectorPaths } style={ style.connectors }></path>
      { React.Children.map(children, (child, i) => (
        <g transform={ childTransforms.get(i) }>
          { React.cloneElement(child, {
            ref: this.childRef(i)
          }) }
        </g>
      ))}
    </React.Fragment>;
  }
}

HorizontalLayout.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  spacing: PropTypes.number,
  withConnectors: PropTypes.bool
};

export default HorizontalLayout;
