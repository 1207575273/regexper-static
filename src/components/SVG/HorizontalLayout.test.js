import React from 'react';
import { mount } from 'enzyme';

import HorizontalLayout from './HorizontalLayout';

import SVGElement from '__mocks__/SVGElement';

describe('HorizontalLayout', () => {
  test('rendering', async () => {
    const component = mount(
      <HorizontalLayout>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </HorizontalLayout>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with connectors', async () => {
    const component = mount(
      <HorizontalLayout withConnectors>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </HorizontalLayout>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });
});
