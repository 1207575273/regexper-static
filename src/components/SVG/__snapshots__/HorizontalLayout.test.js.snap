// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalLayout rendering 1`] = `
<HorizontalLayout
  spacing={10}
  withConnectors={false}
>
  <path
    d=""
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <g
    key=".0"
    transform="translate(0 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
  <g
    key=".1"
    transform="translate(110 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
  <g
    key=".2"
    transform="translate(220 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</HorizontalLayout>
`;

exports[`HorizontalLayout rendering with connectors 1`] = `
<HorizontalLayout
  spacing={10}
  withConnectors={true}
>
  <path
    d="M100,50H110M210,50H220"
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <g
    key=".0"
    transform="translate(0 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
  <g
    key=".1"
    transform="translate(110 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
  <g
    key=".2"
    transform="translate(220 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</HorizontalLayout>
`;
